src/brod_partitions_sync.erl

```diff
@@ -0,0 +1,169 @@
+%%%
+%%%   Copyright (c) 2014-2021 Klarna Bank AB (publ)
+%%%
+%%%   Licensed under the Apache License, Version 2.0 (the "License");
+%%%   you may not use this file except in compliance with the License.
+%%%   You may obtain a copy of the License at
+%%%
+%%%       http://www.apache.org/licenses/LICENSE-2.0
+%%%
+%%%   Unless required by applicable law or agreed to in writing, software
+%%%   distributed under the License is distributed on an "AS IS" BASIS,
+%%%   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%%%   See the License for the specific language governing permissions and
+%%%   limitations under the License.
+%%%
+
+%% @doc A `brod_partitions_sync' is a `gen_server' that is responsible for fetching
+%% the latest partition counts for a client and ensuring external changes to partitions
+%% are propogated to the clients and starts a producer for the partition if not present
+-module(brod_partitions_sync).
+-behaviour(gen_server).
+
+-export([
+    start_link/3
+]).
+
+-export([
+    code_change/3,
+    handle_call/3,
+    handle_cast/2,
+    handle_info/2,
+    init/1,
+    terminate/2
+]).
+
+-include("brod_int.hrl").
+
+-define(DEFAULT_SYNC_PARTITIONS_INTERVAL_SECONDS, 60).
+
+-record(state, {
+    client_id :: pid(),
+    producers_sup :: pid(),
+    interval :: non_neg_integer(),
+    config :: brod_client:config()
+}).
+
+-type state() :: #state{}.
+
+-spec start_link(pid(), pid(), brod_client:config()) ->
+    {ok, pid()} | {error, any()}.
+start_link(ClientId, ProducersSup, Config) ->
+    gen_server:start_link(?MODULE, {ClientId, ProducersSup, Config}, []).
+
+-spec init({pid(), pid(), brod_client:config()}) ->
+    {ok, state()}.
+init({ClientId, ProducersSup, Config}) ->
+    Interval = sync_partition_interval(Config),
+    schedule_sync(Interval),
+    State = #state{
+        client_id = ClientId,
+        producers_sup = ProducersSup,
+        interval = Interval,
+        config = Config
+    },
+
+    {ok, State}.
+
+%% @private
+handle_info(
+    sync,
+    #state{
+        client_id = Client,
+        config = Config,
+        producers_sup = Sup,
+        interval = Interval
+    } = State
+) ->
+    sync_partitions(Client, Sup, Config),
```

Comment:
This ensure the new partitions get a producer but then we need to tell the client to update the metadata so it can use the new partition.